src/webgpu/api/operation/resource_init/buffer.spec.ts

```diff
@@ -1,11 +1,231 @@
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { assert, unreachable } from '../../../../common/util/util.js';
+import { GPUConst } from '../../../constants.js';
 import { GPUTest } from '../../../gpu_test.js';
 
 export const description = `
 Test uninitialized buffers are initialized to zero when read
 (or read-written, e.g. with depth write or atomics).
 
-TODO
+Here lists all the buffer usages to test in future patches:
+- as copy source
+- as copy destination in a partial copy
+- in ResolveQuerySet()
+- as uniform / read-only storage / storage buffer
+- as vertex / index buffer
+- as indirect buffer
 `;
 
-export const g = makeTestGroup(GPUTest);
+const kMapModeOptions = [GPUConst.MapMode.READ, GPUConst.MapMode.WRITE];
+const kBufferUsagesInBufferMappingTests = [
+  GPUConst.BufferUsage.COPY_DST | GPUConst.BufferUsage.MAP_READ,
+  GPUConst.BufferUsage.COPY_SRC | GPUConst.BufferUsage.MAP_WRITE,
+  GPUConst.BufferUsage.COPY_SRC,
+];
+
+class F extends GPUTest {
+  GetBufferUsageFromMapMode(mapMode: GPUMapModeFlags): number {
+    switch (mapMode) {
+      case GPUMapMode.READ:
+        return GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ;
+      case GPUMapMode.WRITE:
+        return GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE;
+      default:
+        unreachable();
+        return 0;
+    }
+  }
+
+  async CheckGPUBufferContent(
+    buffer: GPUBuffer,
+    bufferUsage: GPUBufferUsageFlags,
+    expectedData: Uint8Array
+  ): Promise<void> {
+    // We can only check the buffer contents with t.expectContents() when the buffer usage contains
+    // COPY_SRC.
+    if (bufferUsage & GPUBufferUsage.MAP_READ) {
+      await buffer.mapAsync(GPUMapMode.READ);
+      this.expectBuffer(new Uint8Array(buffer.getMappedRange()), expectedData);
+      buffer.unmap();
+    } else {
+      assert((bufferUsage & GPUBufferUsage.COPY_SRC) !== 0);
+      this.expectContents(buffer, expectedData);
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('partial_write_buffer')
+  .desc(
+    `
+  Verify when we upload data to a part of a buffer with writeBuffer() just after the creation of the
```

Comment:
u-u-nit: I think this will cause an empty line in the test viewer. I'm not sure what's the best practice for desc messages.

Reply:
ditto below of course